### ISSUE TITLE
shutil.make_archive should not need to chdir (alternatively: make shutil.make_archive thread-safe)

### ISSUE THREAD

**Issue by 030a70bd-87f0-44f1-bce5-e413ed5608b3** [python/cpython#74696]
BPO | [30511](https://bugs.python.org/issue30511)
--- | :---
Nosy | @tarekziade, @ambv, @serhiy-storchaka, @miss-islington, @michael-o, @FFY00, @akulakov
PRs | <li>python/cpython#26933</li><li>python/cpython#27274</li><li>python/cpython#27275</li><li>python/cpython#27276</li><li>python/cpython#28271</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-05-30.13:46:18.840>
labels = ['3.7', 'library']
title = 'shutil.make_archive should not need to chdir (alternatively: make shutil.make_archive thread-safe)'
updated_at = <Date 2021-09-17.13:26:13.713>
user = 'https://bugs.python.org/AlexGaynor'
```

bugs.python.org fields:
```python
activity = <Date 2021-09-17.13:26:13.713>
actor = 'FFY00'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-05-30.13:46:18.840>
creator = 'Alex Gaynor'
dependencies = []
files = []
hgrepos = []
issue_num = 30511
keywords = ['patch']
message_count = 10.0
messages = ['294756', '294758', '294759', '294760', '297902', '389849', '396633', '397928', '397932', '397934', '397935']
nosy_count = 9.0
nosy_names = ['tarek', 'lukasz.langa', 'serhiy.storchaka', 'Alex Gaynor', 'Joey Harrington', 'miss-islington', 'michael-o', 'FFY00', 'andrei.avk']
pr_nums = ['26933', '27274', '27275', '27276', '28271']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue30511'
versions = ['Python 2.7', 'Python 3.7']
```

</p></details>

**Comment by 030a70bd-87f0-44f1-bce5-e413ed5608b3** [comment]
Currently shutil.make_archive uses os.chdir, however there's no need for that. Everything that's done could be equally accomplished with path manipulation:

https://github.com/python/cpython/blob/master/Lib/shutil.py#L773-L779

We should switch to using path manipulation in order to make shutil.make_archive thread safe.

(Flag: This is probably a good bug for someone with Python skills interested in contributing to CPython!)

**Comment by serhiy-storchaka** [comment]
Unfortunately this is not possible. The signature of functions registered with register_archive_format() doesn't support this.

**Comment by 030a70bd-87f0-44f1-bce5-e413ed5608b3** [comment]
None of those functions are a public API, so changing them shouldn't be a problem IMO.

**Comment by 030a70bd-87f0-44f1-bce5-e413ed5608b3** [comment]
Ugh, except via |register_archive_format|. |register_archive_format| could wrap callables passed to it to maintain the current behavior.

**Comment by db51f3f2-c18a-47f4-b66d-30328ae2ee48** [comment]
It would be nice if there was at least a warning in the docs that make_archive is not thread-safe, and that if you have two threads creating archives that it's extremely likely you'll get erroneous results since the race condition lasts for the entire duration of the archive creation.

**Comment by gpshead** [comment]
Just a note that this is still relevant. We still get Python users at work who run into this. _(the documentation definitely helps, but often after the fact while debugging a WTF?! style failure)_

I wouldn't let `register_archive_format()` et. al. hold us back. Such APIs can be deprecated with a new one requiring people to provide a thread safe archiver.

We can detect when we're using an unknown externally registered format that is not chdir safe (anything unknown that came to us via `register_archive_format()`). We could add a `make_archive()` bool keyword argument `thread_safe=False` as a 3.12 feature to make attempts to use such an archiver raise an exception when True. We'd want to cleanup our existing built-in archiver functions to be thread safe ones as part of this.

**Comment by gpshead** [comment]
Also a reality check: What code exists that uses `register_archive_format()` in the wild? I find no uses in our huge internal codebase that included copies of many pypi/github OSS things.  _(surprising, usually there's at least 1 for anything no matter how esoteric it seems)_ This could be a very smooth deprecation process given so few apparent users...

**Comment by serhiy-storchaka** [comment]
Okay, I take this.

**Comment by serhiy-storchaka** [comment]
#93160 is a first step. It makes standard zip and tar archivers not changing the current working directory. In the following PR I'll make it supporting in custom archivers.

**Comment by ambv** [comment]
@gpshead I'm +1 to deprecate `register_archive_format()` in 3.12 and remove in 𝛑-Py. Can we use this issue for that or should I open a new one?

**Comment by serhiy-storchaka** [comment]
I was going to extend `register_archive_format()` by adding support of custom unpackers which support `root_dir`.

It can be implemented either by adding a boolean parameter `supports_root_dir` in `register_archive_format()`, or checking the `supports_root_dir` attribute of the unpacker function.

**Comment by ambv** [comment]
Sure, Serhiy. Greg's point is that this functionality looks unused so it's probably an unnecessary support burden.

**Comment by serhiy-storchaka** [comment]
[py7zr](https://pypi.org/project/py7zr/) provides helpers to be registered in `shutil`. Other archiver modules can provide such helpers too or even register them by default. Removing `register_archive_format()` and `register_unpack_format()` will make other archive formats second-class citizens.

**Comment by serhiy-storchaka** [comment]
I have implemented the second option in #94251. It seems better because does not need a collaboration between the author and users of the custom archiver.
